### PR TITLE
Implement Cancel Methods for Sql Engines

### PR DESCRIPTION
**Pull request by plypaul**
This PR implements the query cancellation methods for the SQL engines (minus Databricks).

### CODE REVIEW

File: metricflow/sql_clients/snowflake.py [hunk 1]
```diff
@@ -244,4 +245,25 @@ def cancel_submitted_queries(self) -> None:  # noqa: D
                     conn.execute(f"SELECT SYSTEM$cancel_all_queries({session_id})")
 
     def cancel_request(self, pattern_tag_set: SqlRequestTagSet) -> int:  # noqa: D
-        raise NotImplementedError
+        # Running queries have an end_time set to the epoch time:
```
Comment (tlento) [inline metricflow/sql_clients/snowflake.py:248]: What? So many interesting choices in the world....

File: metricflow/sql_clients/big_query.py [hunk 2]
```diff
@@ -57,29 +67,39 @@ def from_connection_details(url: str, password: Optional[str]) -> BigQuerySqlCli
...
-        super().__init__(
-            engine=BigQuerySqlClient._create_bq_engine(
-                query_string=query_string, project_id=project_id, password=password
-            )
+        password_json = json.loads(password) if password else {}
+        self._project_id = project_id or password_json.get("project_id")
+        bq_engine = BigQuerySqlClient._create_bq_engine(project_id=project_id, password=password)
+        self._bq_client = bigquery.Client(
+            project=self._project_id,
+            credentials=google.oauth2.service_account.Credentials.from_service_account_info(password_json),
```
Comment (tlento) [inline metricflow/sql_clients/big_query.py:80]: Why are all of these changes in this commit?

Also, does this work if we aren't using service accounts?

File: metricflow/sql_clients/big_query.py [hunk 3]
```diff
@@ -90,10 +110,15 @@ def _create_bq_engine(
            - Google's decision tree (https://google.aip.dev/auth/4110)
 
         Args:
-            query_string: Additional query params to pass to engine construction.
+            query_field_values: The field/value pairs to use in the query string.
```
Comment (tlento) [inline metricflow/sql_clients/big_query.py:113]: This I get. That was super weird with the bare string.

File: metricflow/sql_clients/big_query.py [hunk 4]
```diff
@@ -57,29 +67,39 @@ def from_connection_details(url: str, password: Optional[str]) -> BigQuerySqlCli
...
+            query_field_values={"dry_run": "true"}, project_id=self._project_id, password=self._password
         )
         with _engine.connect() as conn:
             conn.execute(sqlalchemy.text(stmt), bind_params.param_dict)
 
+    @staticmethod
+    def _create_bq_url(project_id: str = "", query_field_values: Optional[Dict[str, str]] = None) -> str:
+        if query_field_values is None:
+            query_field_values = {}
+        query_items = tuple(f"{field}:{value}" for field, value in query_field_values.items())
```
Comment (tlento) [inline metricflow/sql_clients/big_query.py:96]: Wait, `{field}:{value}` not `{field}={value}`?

File: metricflow/sql_clients/big_query.py [hunk 5]
```diff
@@ -90,10 +110,15 @@ def _create_bq_engine(
...
 
         Args:
-            query_string: Additional query params to pass to engine construction.
+            query_field_values: The field/value pairs to use in the query string.
             project_id: Project ID listed on the GCP platform.
             password: String representation of keyfile.json.
         """
+        if query_field_values is None:
+            query_field_values = {}
+        query_items = tuple(f"{field}:{value}" for field, value in query_field_values.items())
```
Comment (tlento) [inline metricflow/sql_clients/big_query.py:119]: Again, why `:` instead of `=` between param key and value?

**Review by tlento — COMMENTED** [review]
The BigQuery change is weirdly expansive and not all of it makes sense to me. Mind providing some more context, or splitting out the other bits if appropriate?

**Review by tlento — DISMISSED** [review]
Hey can you look over the inlines? I don't fully understand all of the changes in here, and I'm not 100% sure they're all correct, but if you've verified and this was what had to happen just some explanation is enough.

File: metricflow/sql_clients/big_query.py (hunk 2 above)
Comment (plypaul) [reply metricflow/sql_clients/big_query.py:80]: Yeah, some of these changes could have been a separate PR, but saw some things that could use fixing.

Not entirely sure about initialization without credentials, but this is the way that the docs create a client: https://github.com/googleapis/python-bigquery/blob/HEAD/samples/client_query.py

File: metricflow/sql_clients/big_query.py (hunk 4 above)
Comment (plypaul) [reply metricflow/sql_clients/big_query.py:96]: That's a mistake - I was looking at the syntax for labels in the URL. Fixed.

File: metricflow/sql_clients/big_query.py (hunk 5 above)
Comment (plypaul) [reply metricflow/sql_clients/big_query.py:119]: Same as previous comment.

File: metricflow/sql_clients/big_query.py [hunk 6]
```diff
@@ -57,29 +67,33 @@ def from_connection_details(url: str, password: Optional[str]) -> BigQuerySqlCli
...
-        super().__init__(
-            engine=BigQuerySqlClient._create_bq_engine(
-                query_string=query_string, project_id=project_id, password=password
-            )
+        password_json = json.loads(password) if password else {}
+        self._project_id = project_id or password_json.get("project_id")
+        bq_engine = BigQuerySqlClient._create_bq_engine(project_id=project_id, password=password)
+        self._bq_client = bigquery.Client(
+            project=self._project_id,
+            credentials=google.oauth2.service_account.Credentials.from_service_account_info(password_json)
```
Comment (tlento) [inline metricflow/sql_clients/big_query.py:80]: Still a little worried about this one, and I'd prefer it be separate just so we don't get confused with the canceling vs the connection change, but I guess it's small enough we can figure it out.

**Review by tlento — APPROVED** [review]
Ok!

**Pull request merged**